bin/build-iso

```diff
@@ -0,0 +1,187 @@
+bi_rootdir=""
+bi_protoboot=${BI_PROTOBOOT}
+bi_iso_download=""
+
+function cleanup()
+{
+	print "Cleaning up"
+
+	if [[ "$bi_iso_download" != "" ]]; then
+		pfrun umount $bi_tmpdir/smartos-iso
+	fi
+
+	pfrun /bin/rm -rf $bi_tmpdir
+}
+
+#
+# Construct the EFI System Partition (ESP) image,  We size it at 256 MB, which
+# is intentionally much larger than what we need currently, in order to leave
+# headroom for future projects which may need to store data in the ESP.
+#
+function create_esp()
+{
+	local readonly tmpdir=$1
+	local readonly efimnt=$2
+	local readonly esp_size=256 # MiB
+	local readonly esp_sects=$(( $esp_size * 1024 * 1024 / 512 ))
+
+	pfrun mkfile -n ${esp_size}m $tmpdir/esp.img
+	bi_esp_blkdev=$(pfexec lofiadm -a $tmpdir/esp.img)
+	if [[ $? != 0 ]]; then
+		pfrun rm $tmpdir/esp.img
+		fail "Failed to create ESP lofi device"
+	fi
+	readonly esp_rawdev=${bi_esp_blkdev/lofi/rlofi}
+
+	pfrun mkfs -F pcfs -o b=system,size=$esp_sects,nofdisk,fat=32 \
+	    $esp_rawdev </dev/null
+
+	pfrun mkdir -p $efimnt
+	pfrun mount -F pcfs -o foldcase $bi_esp_blkdev $efimnt
+	pfrun mkdir -p $efimnt/efi/boot
+	pfrun cp $bi_protoboot/loader64.efi $efimnt/efi/boot/bootx64.efi
+	pfrun umount $efimnt
+	pfrun lofiadm -d $bi_esp_blkdev
+}
+
+# - Populate root from the bi_boot_image.
+function populate_root()
+{
+	local readonly dir=$1
+
+	print "Installing boot tarball onto root partition ... \c"
+	pfrun gtar -xzf "$bi_path" -C $dir
+
+	## Make any bootparams adjustments here.
+	cp $dir/boot/loader.conf /tmp/loader.conf.$$
+	sed 's/ipxe="true"/ipxe="false"/g' < /tmp/loader.conf.$$ > \
+	    $dir/boot/loader.conf
+	rm -f /tmp/loader.conf.$$
+	echo 'triton_installer="iso"' >> $dir/boot/loader.conf
+
+	# XXX e.g. more bootparams?!?
+
+	# chmod EVERYTHING to be executable like on the USB key.
+	pfrun chmod -R u+x $dir
+	# And while I'm at it, pfexec chown to root.
+	pfrun chown -R root $dir
+
+	print "done"
+}
+
+function create_iso()
+{
+	local readonly populate=$1
+	local readonly iso=$2
+	local readonly espimg=$3
+	local readonly 
+
+	if [[ "$populate" == "yes" ]]; then
+	    pfrun mkdir -p $bi_rootdir
+	    populate_root $bi_rootdir
+	fi
+
+	pfrun cp $bi_protoboot/cdboot $bi_rootdir/boot/cdboot
+	pfrun cp $espimg $bi_rootdir/boot/efiboot.img
+
+	pfrun mkisofs -quiet -R \
+	    -eltorito-boot boot/cdboot -no-emul-boot -boot-info-table \
+	    -eltorito-alt-boot -eltorito-platform efi \
+	    -eltorito-boot boot/efiboot.img -no-emul-boot \
+	    -o $iso $bi_rootdir
+
+	pfrun chown -R ${uid}:${gid} $iso
+
+	print "Successfully created $iso"
+}
+
+bi_tmpdir=$(mktemp -d -p /var/tmp) || fail "mktemp failed!"
+
+trap cleanup EXIT
+bi_efimnt=$bi_tmpdir/mnt/efi
+bi_rootdir=$bi_tmpdir/mnt/root
+iso_image_name=iso-${bi_boot_image_name}.iso
+SMARTOS_URL=${SMARTOS_URL:-https://us-east.manta.joyent.com/Joyent_Dev/public/SmartOS/smartos-latest.iso}
+
+if [[ "$bi_protoboot" == "" ]]; then
+	# We need to download a standlone SmartOS ISO to get "cdboot".
```

Comment:
Is this likely to happen frequently?

Reply:
Only if you `gmake ipxe` or `gmake iso`.  Also, if you set BI_PROTOBOOT for a built smartos-live, it'll not do that and instead copy straight from there.

Reply:
Ok, so not as part of our standard build, we'll just reuse the artifacts we built earlier.